proto/substrait/extended_expression.proto

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: Apache-2.0
+syntax = "proto3";
+
+package substrait;
+
+import "substrait/algebra.proto";
+import "substrait/extensions/extensions.proto";
+import "substrait/plan.proto";
+import "substrait/type.proto";
+
+option csharp_namespace = "Substrait.Protobuf";
+option go_package = "github.com/substrait-io/substrait-go/proto";
+option java_multiple_files = true;
+option java_package = "io.substrait.proto";
+
+message ExpressionReference {
+  oneof expr_type {
+    Expression expression = 1;
+    AggregateFunction measure = 2;
+  }
+  // Field names in depth-first order
+  repeated string output_names = 3;
```

Comment:
It's strange to me that `output_names` are in `ExpressionReference` and not `ExtendedExpression` but this is a natural consequence of have repeated expressions at the top level.

Reply:
I have thought about this two options. Both of them can work as expected but seems it makes things complicated if we put it  in `ExtendedExpression`. Think about 2 expressions that one has a single output column called `output_a` and the other has 2 output columns called `output_b, output_c`. If we put it in  `ExtendedExpression`, then the names should like `output_a, output_b, output_c`, then we have to do a mapping job when try to connect the column to its name. Do you think so?

Reply:
I think I am confused.  How can an expression have more than one output column?

Reply:
I am thinking about a function like `SELECT CAST(JSON '{"k1":1,"k2":23,"k3":456}' AS MAP(VARCHAR, INTEGER));` It should have 2 columns output in a columnar computation scenario.

Reply:
Yes, that would have multiple output_names, I see now.

> Think about 2 expressions that one has a single output column called output_a and the other has 2 output columns called output_b, output_c.

Only one of these should be the top-level expression correct?  That is the only one whose names we care about.

For example, if we have three expressions:

```
expr_ref(0) + expr_ref(1): output_names="final"
field(0) * expr_ref(2): output_names="middle"
field(1) * 20: output_names="first"
```

The only meaningful value of `output_names` is "final" correct?

Reply:
Is there any special meaning between names "first", "middle" and "final"? I mean is there any implicit relation between these expressions? if not, in case that they equally exist in a project node, which has three output columns for following plan node, all of these names are meaningful and necessary. Following plan node probably refer the columns by these names. Example pattern likes `agg (sum("a"), min("b"), max("c") <- project(expr_0:"a", expr_1:"b", expr_2:"c")`.

Reply:
I see your point.  Sorry, I was thinking there was only one top-level expression.